### PR TITLE
fix: IPC on multiwindowed linux apps (fix #11171)

### LINKED ISSUE #11171: [bug] broken IPC in multi window app on Linux
### Describe the bug

In a multi-window Tauri app on Linux, only the first window ever created can receive events and have its `listen` callbacks triggered. All subsequent windows do not receive any events.

In this video, the same Tauri app runs on Windows on the left and Linux on the right.

Every window ran `listen("reply", (event) => {alert(event.payload})})`.
When the button is pressed, a Rust Tauri command is invoked, which emits the `reply` event.

On Windows, the app behaves correctly and every window receives the `reply` event.

On Linux, only the first window ever receives any events.

https://github.com/user-attachments/assets/b2c0fd72-d7b8-4764-a92d-f82a2dd9af7b

### Reproduction

The example shown in the video: https://github.com/Ynng/tauri-listen-test

Or:

- Clone the tauri repo on an Ubuntu 24 machine
- Run `cargo run --example multiwindow`
- Try to send a message to the second/third window

### Expected behavior

All windows should receive events that are "emitted to all targets".

### Full `tauri info` output

```text
[✔] Environment
    - OS: Fedora 40.0.0 x86_64 (X64)
    ✔ webkit2gtk-4.1: 2.44.3
    ✔ rsvg2: 2.57.1
    ✔ rustc: 1.80.1 (3f5fd8dd4 2024-08-06)
    ✔ cargo: 1.80.1 (376290515 2024-07-16)
    ✔ rustup: 1.27.1 (54dd3d00f 2024-04-24)
    ✔ Rust toolchain: stable-x86_64-unknown-linux-gnu (environment override by RUSTUP_TOOLCHAIN)
    - node: 21.7.2
    - pnpm: 9.7.1
    - npm: 10.5.0

[-] Packages
    - tauri 🦀: 2.0.0-rc.11
    - tauri-build 🦀: 2.0.0-rc.13
    - wry 🦀: 0.44.1
    - tao 🦀: 0.30.2
    - tauri-cli 🦀: 2.0.0-rc.10

[-] Plugins
    - tauri-plugin-shell 🦀: 2.0.0-rc.0

[-] App
    - build-type: bundle
    - CSP: unset
    - frontendDist: ../dist
    - devUrl: http://localhost:1420/
```

### Stack trace

_No response_

### Additional context

Undoing #11043 fixes this problem, but re-introduces #10981

### PR DESCRIPTION
**Pull request by Ynng**
closes #11171

- Fix ipc on multiwindowed linux apps
- Include window label in `InvokeRequest`
- Update the IPC handler to fetch webview based on the window label in `InvokeRequest`

### CODE REVIEW

**Comment by amrbashir** [comment]
@Ynng thanks for investigating and making a PR, I wonder if we can fix this in `wry` instead.

**Comment by Ynng** [comment]
> @Ynng thanks for investigating and making a PR, I wonder if we can fix this in `wry` instead.

As in fixing the underlying "Cannot register URI scheme xxx more than once"?

**Comment by amrbashir** [comment]
thanks @Ynng, I have decided to incorporate the idea of this PR in wry instead, see https://github.com/tauri-apps/wry/pull/1384 and https://github.com/tauri-apps/tauri/pull/11268. Apologies for the inconvenience.

@lucasfernog maybe we should merge this as a fix for 2.0.0 because #11268 will require a minor bump.

File: crates/tauri/src/webview/mod.rs [hunk 1]
```diff
@@ -132,6 +132,8 @@ pub struct InvokeRequest {
   pub headers: HeaderMap,
   /// The invoke key. Must match what was passed to the app manager.
   pub invoke_key: String,
+  /// The label of the webview that requested this command.
+  pub label: String,
```
Comment (amrbashir) [inline crates/tauri/src/webview/mod.rs:136]: if we are gonna merge this PR as temporary fix, we should avoid adding this field.

File: crates/tauri/src/ipc/protocol.rs [hunk 2]
```diff
@@ -532,6 +531,14 @@ fn parse_invoke_request<R: Runtime>(
       .map_err(|_| "Tauri error header value must be a numeric string")?,
   );
 
+  let label = parts
+    .headers
+    .get(TAURI_WEBVIEW_LABEL_HEADER_NAME)
+    .ok_or("missing Tauri-Webview-Label header")?
```
Comment (amrbashir) [inline crates/tauri/src/ipc/protocol.rs:537]: could you also fix tests that are failing because of this?

**Review by amrbashir — CHANGES_REQUESTED** [review]

**Comment by Ynng** [comment]
Thanks @amrbashir for the proper fix in wry (tauri-apps/wry#1384 and #11268)!

Given that this PR introduces temporary changes across multiple files and a new field in `InvokeRequest`, I'm okay with closing it in favor of waiting for the wry-based solution.

**Comment by amrbashir** [comment]
@Ynng This PR can still be merged to fix the current released minor, we just need the changes I requested

**Comment by amrbashir** [comment]
thanks @Ynng , I will go ahead and close this then since https://github.com/tauri-apps/tauri/pull/11268 will be merged and released as patch

**Pull request closed without merging**